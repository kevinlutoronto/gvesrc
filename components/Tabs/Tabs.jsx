// @flow
import React from 'react';
import cn from 'classnames';
import { map } from 'lodash';
import { getDataAttrs } from '../../dataUtils';
import type { Data } from '../../types';
import style from './style.scss';

type Tab = {|
  label: string,
  title: React$Node,
|};

type Props = {|
  items: Array<Tab>,
  onChange?: (string, number) => void,
  selectedIndex?: number,
  selectedLabel?: string,
  data?: Data,
  plain?: string,
  vertical?: boolean,
|};

class Tabs extends React.Component<Props> {
  static defaultProps = {
    items: [],
    onChange: () => {},
  };

  onClick = (tab: Tab, index: number): void => {
    this.props.onChange && this.props.onChange(tab.label, index);
  };

  onKeyDown = (event: { keyCode: number }, tab: Tab, index: number): void => {
    if (event.keyCode === 13 /* enter */) {
      this.onClick(tab, index);
    }
  };

  getTabsStyle(): string {
    const { vertical, plain } = this.props;
    if (vertical) return style.verticalTabs;
    if (plain) return style.plainTabs;
    return style.tabs;
  }

  getTabStyle(): string {
    const { vertical, plain } = this.props;
    if (vertical) return style.verticalTab;
    if (plain) return style.plainTab;
    return style.tab;
  }

  getTabClassNames = (tab: Tab, index: number): string => {
    const baseTabClass = this.getTabStyle();
    return cn(baseTabClass, {
      [style.active]: this.props.selectedIndex === index || this.props.selectedLabel === tab.label,
      [style.first]: index === 0,
      [style.last]: index === this.props.items.length - 1,
    });
  };

  renderTab = (tab: Tab, index: number) => {
    return (
      <li
        role="tab"
        onClick={() => this.onClick(tab, index)}
        onKeyDown={event => this.onKeyDown(event, tab, index)}
        className={this.getTabClassNames(tab, index)}
        key={index}
        tabIndex="0"
      >
        {tab.title}
      </li>
    );
  };

  render() {
    return (
      <ul {...getDataAttrs(this.props.data)} className={this.getTabsStyle()} role="tablist">
        {map(this.props.items, this.renderTab)}
      </ul>
    );
  }
}

Tabs.displayName = 'Plasma@Tabs';

export default Tabs;
