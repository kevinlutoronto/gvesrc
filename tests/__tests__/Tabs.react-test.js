/* eslint-disable */

import React from 'react';
import Tabs from '../../components/Tabs/Tabs';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('Tabs Component', () => {
  test('render', () => {
    const component = renderer.create(<Tabs />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('two items', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
      {
        label: 'two',
        title: 'Two',
      },
    ];
    const component = shallow(<Tabs items={items} />);
    expect(component.find('li')).toHaveLength(2);
  });

  test('onClick first tab', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
      {
        label: 'two',
        title: 'Two',
      },
    ];
    const onChange = jest.fn();
    const component = mount(<Tabs items={items} onChange={onChange} />);
    component
      .find('li')
      .first()
      .simulate('click');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('one', 0);
  });

  test('onClick last tab', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
      {
        label: 'two',
        title: 'Two',
      },
    ];
    const onChange = jest.fn();
    const component = mount(<Tabs items={items} onChange={onChange} />);
    component
      .find('li')
      .last()
      .simulate('click');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('two', 1);
  });

  test('keydown first tab', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
      {
        label: 'two',
        title: 'Two',
      },
      {
        label: 'three',
        title: 'Three',
      },
    ];
    const onChange = jest.fn();
    const component = mount(<Tabs items={items} onChange={onChange} />);
    component
      .find('li')
      .first()
      .simulate('keydown', { keyCode: 13 });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('one', 0);
  });

  test('keydown last tab', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
      {
        label: 'two',
        title: 'Two',
      },
      {
        label: 'three',
        title: 'Three',
      },
    ];
    const onChange = jest.fn();
    const component = mount(<Tabs items={items} onChange={onChange} />);
    component
      .find('li')
      .last()
      .simulate('keydown', { keyCode: 13 });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('three', 2);
  });
});
