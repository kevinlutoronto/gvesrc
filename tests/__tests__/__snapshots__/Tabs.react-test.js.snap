// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component render 1`] = `
<ul
  className="tabs"
  role="tablist"
/>
`;
